### DISCUSSION TITLE
How to POST an `Accept` activity to Mastodon follower's inbox after a `Follow`?

### DISCUSSION THREAD

**Discussion by markcellus** [mastodon/mastodon#18453] [Q&A] [answered]
Hello, I've been working on a Fediverse client + server written only in NodeJS and would like for it to be interoperable with Mastodon. I've got a user of the client showing up in Mastodon search, which works great! And now I'm working on adding the ability for Mastodon users to follow a user from my server. 

I've implemented everything on [How to Make Friends and Verify Requests](https://blog.joinmastodon.org/2018/07/how-to-make-friends-and-verify-requests/) doc page to sign requests and implemented the `Digest` header as mentioned in [Digest header / federation](https://discourse.joinmastodon.org/t/digest-header-federation/3328).

But when my server attempts to POST an `Accept` (follow) activity to the Mastodon user's inbox to confirm their `Follow`, I'm getting the following error (replaced real domain with my-example.com for privacy):

```
Verification failed for bob@my-example.com https://my-example.com/actor using rsa-sha256 (RSASSA-PKCS1-v1_5 with SHA-256)
```

I'm implementing the `rsa-sha256` hash creation using [`createHash` of NodeJS's crypto module](https://nodejs.org/api/crypto.html#cryptocreatehashalgorithm-options) for `Digest` and the [`sign` method for the HTTP Signature](https://nodejs.org/api/crypto.html#cryptosignalgorithm-data-key-callback). Could it be possible that the NodeJS's algorithms may output different values from Ruby's implementation? Any help would be appreciated. Thanks!

**Comment by nightpool** [comment]
Can you give an example of the string you're signing and the code you used to generate it? Probably you're putting your parameters in the wrong order or not canonicalizing properly.

**Reply by markcellus** [reply]
Sure. But first, let's verify a few assumptions I'm making...

1. The [HTTP Signatures](https://blog.joinmastodon.org/2018/06/how-to-implement-a-basic-activitypub-server/#http-signatures) section doesn't mention anything about `Digest` header, but it seems to be an additional header. So I've added it to `signed_string` before signing. But maybe that is incorrect. Can you confirm whether the `Digest` header I've generated is supposed to be included in the `signed_string` mentioned on that page?  
2. When my server makes the POST request to the Mastodon user's inbox, there are other headers that can be automatically sent outside of the ones concatenated onto `signed_string` (i.e. `User-Agent`, `DNT`, etc). Should I also be adding those to the signature? Or is just signing with `signed_string` set to `(request-target): post /inbox host date digest` sufficient?

The answers to my questions above may get me closer to a more accurate implementation. If that doesn't help, I'll write up an example for you. Thank you so much for the reply and the help! :+1:

**Reply by ClearlyClaire** [reply]
The `signed_string` should look like something like:
```
(request-target): post /inbox
host: domain.org
date: 2022-…
digest: SHA-256=…
```

and the `Signature` header should look like something like: `keyId="https://my-example.com/actor",headers="(request-target) host date digest",signature=<signature>`

**Comment by markcellus** [comment, marked as answer]
Ah, thanks @ClearlyClaire and @nightpool for the help. Turns out I wasn't sending the same stringified JSON object as the one passed as the body in the request. So it was causing an issue. But everything works now. Thanks!

**Reply by sventiffe** [reply]
Hi Mark, I am running into the same issue (` Verification failed for...`). I'm running into this before even sending the JSON payload, see detailed description [here](https://github.com/mastodon/mastodon/discussions/21561). Is there anything that you can see in my post that could help me troubleshooting it?

**Reply by sventiffe** [reply]
After initial investigation it seems that my problem throws the same error message but is unrelated. I've also realized that it does not make much sense to follow up on a discussion that is marked as "answered", hence posted a [separate question](https://github.com/mastodon/mastodon/discussions/21621). Sorry for having spammed this resolved discussion.
